Deprecate single product block save

octaedro opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR fixes an error introduced in PR #51023. We changed the static block `Single Product` without using the deprecation API. As a result, an error was displayed. There is a GH discussion [here](https://github.com/woocommerce/woocommerce/discussions/51098), to continue analyzing a solution for the WooCommerce pages styles problem.

![image](https://github.com/user-attachments/assets/5e49c2c7-131a-404e-b012-f8e331099fe2)

This PR also includes a minor fix to prevent applying some WooCommerce styles to the editor button inside the `Single Product` block.

**Before** (note that the `Add to cart` button is violet, but it shouldn't have that color)

![Screenshot 2024-09-04 at 11 23 14 AM](https://github.com/user-attachments/assets/1e822aa3-3d0a-45ad-9d92-0961120195bb)

**After** (after this PR the button looks correct)

![Screenshot 2024-09-04 at 11 21 25 AM](https://github.com/user-attachments/assets/1f6c781c-d613-49a1-aff4-33a8f80672a1)

 
<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Go to Products > Add New and create a variable product with a few variations.
2. Checkout a commit older than [3def186](https://github.com/woocommerce/woocommerce/commit/3def18623e9b2a8df170000f6f3dbfb07b8caab7), for instance: `dec89a13111f5c8cf2bc6864677b63453bc0ef35`.
Run the following command in your terminal:
```
git checkout dec89a13111f5c8cf2bc6864677b63453bc0ef35
```
3. Build the project and go to Pages > Add New Page
4. Add a `Single Product` block and select the variable product you just created (do not select a variation, because we want to see the select used to change from one variation to another).

![Screenshot 2024-09-04 at 11 40 38 AM](https://github.com/user-attachments/assets/f8a6ef9e-30f7-47cd-ae67-969cdccc8dbc)

5. Save the page.
6. Checkout `trunk` and rebuild the project.
7. You will see an error because the block is not passing the validation since the block changed [here](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/assets/js/blocks/single-product/save.tsx#L7-L9).

![Screenshot 2024-09-04 at 11 44 28 AM](https://github.com/user-attachments/assets/55650ec6-a3d7-46b9-8411-db1f9ce7373c)

8. Now add another `Single Product` block to the page. Select the same variable product you created in step 1.
9. Update your page and view it.
10. You will notice the first product has the selector's styles broken and the second one appears fine. That happens because the second one has the class `woocommerce` that fixes the selector styles.

![Screenshot 2024-09-04 at 11 51 59 AM](https://github.com/user-attachments/assets/8fd1274d-527e-4688-8a28-7c0ede4b70ef)

11. Now checkout this branch and build the project.
```
git checkout fix/deprectated_single_product_block_save
```
12. Go back to the post editor and refresh it.
13. You won't see the error mentioned in step 7 anymore.

This PR fixes the validation error and a styling bug. To reproduce the bug you will have to: 

14. Checkout the old commit again and build your project using the following command:
```
git checkout dec89a13111f5c8cf2bc6864677b63453bc0ef35
```
15. Refresh the page editor you were editing.
16. The `Add to cart` button is now violet (the button styles should be different).

![Screenshot 2024-09-04 at 11 23 14 AM](https://github.com/user-attachments/assets/1e822aa3-3d0a-45ad-9d92-0961120195bb)

17. After checking out this branch and rebuilding the project, the styling bug shouldn't be there anymore.

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

octaedro committed:
Add deprecated save

octaedro committed:
Fix button styles

octaedro committed:
Add changelog

tjcafferkey commented:
Just a couple of things from me. Firstly I think we should use this PR to put a comment [here](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/assets/js/blocks/single-product/save.tsx#L7-L9) to explain why we are doing this. Future us will appreciate that.

Secondly, when I build this branch and go back to the edit post page to view my original block (step 12) I get the "no product" view of the block. I went through the testing steps twice with different products each time and got the same result. Can you confirm if you are experiencing the same?

![Screenshot 2024-09-05 at 08 28 37](https://github.com/user-attachments/assets/37539bf5-1f03-42f2-86d3-f59b1462071f)

octaedro committed:
Add attributes to deprecated prop

octaedro committed:
Add supports prop to deprecated obj

octaedro committed:
Add code comment

octaedro commented:
Thank you @tjcafferkey, for your comments,

I just addressed your suggestions.

> Just a couple of things from me. Firstly I think we should use this PR to put a comment [here](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/assets/js/blocks/single-product/save.tsx#L7-L9) to explain why we are doing this. Future us will appreciate that.

Done in [f900cd2](https://github.com/woocommerce/woocommerce/pull/51153/commits/f900cd20b250f9f8d47840b2eb3038e69ae1e150)

> Secondly, when I build this branch and go back to the edit post page to view my original block (step 12) I get the "no product" view of the block. I went through the testing steps twice with different products each time and got the same result. Can you confirm if you are experiencing the same?

Done in commits [f848654](https://github.com/woocommerce/woocommerce/pull/51153/commits/f84865458b654449174d88674791b470fab6650c) and [c1aac7f](https://github.com/woocommerce/woocommerce/pull/51153/commits/c1aac7fb28a098395dba0dda7e7da608b3133aa4)

louwie17 commented on plugins/woocommerce/client/legacy/css/woocommerce.scss:
```diff
@@ -1699,7 +1699,10 @@ p.demo_store,
 .woocommerce:where(body:not(.woocommerce-block-theme-has-button-styles)),
-:where(body:not(.woocommerce-block-theme-has-button-styles)) .woocommerce {
+:where(body:not(.woocommerce-block-theme-has-button-styles)):where(
+		:not(.edit-post-visual-editor *)
```
I am confused by this change, this removes any styling from the buttons within the editor, you would think we want that?
Do you have a reference to an issue or discussion that suggested we shouldn't do this?

octaedro replied:
This code addresses the second point I mentioned in the PR description: the `Add to cart` button changes its color after the `woocommerce` class is added to the block in the editor.

The scope of those styles wasn’t specific enough, causing some frontend styles to be unintentionally applied to backend blocks in the editor. That’s why I added the `:not(.edit-post-visual-editor *)` to ensure proper scoping.

louwie17 replied:
Gotcha, makes sense than, although it also slightly weirds me out. As I thought the site editor should depict the frontend styles, as it is a visual editing experience.

octaedro replied:
> As I thought the site editor should depict the frontend styles, as it is a visual editing experience.

That’s not happening at the moment 😕, but it might be a good idea to explore depicting the frontend styles in the editor. I do wonder if that would affect or potentially break the editing experience.

louwie17 commented on plugins/woocommerce-blocks/assets/js/blocks/single-product/deprecated.tsx:
```diff
@@ -0,0 +1,28 @@
+
+const v1 = {
+	attributes: metadata.attributes,
+	supports: metadata.supports,
```
Nice!

louwie17 reviewed (commented):
This tested well, although I did have one question around the changes in the CSS file, as I can't find any specific reference for that change.
Also I battled with the testing instructions a bit, as some of them were unclear:

> Add a Single Product block and select a product.

This was unclear if I should be selecting `any` product or the variable product I just created, or a variation of the product I just created. Turns out it was the variable product and not the variation, it may be helpful to mention this, as the UI makes it also look like that your forced to select a `variation` although the variable product is also an option.

> The buttons are now violet

Maybe explicitly say that the buttons should be violet in the page editor, as I was confused why they weren't violet within the preview, but only in the editor context.

octaedro commented:
Hi @louwie17, thank you for reviewing this PR.

> This tested well, although I did have one question around the changes in the CSS file, as I can't find any specific reference for that change.

The changes to the CSS file address the second point mentioned in the PR description: _a minor fix to prevent applying some WooCommerce styles to the editor button inside the Single Product block_.
After adding the `woocommerce` class to the `Single Product` block, some frontend styles were unintentionally applied to the `Single Product` block in the editor. I added the `:not(.edit-post-visual-editor *)` to ensure proper scoping.

> Also I battled with the testing instructions a bit, as some of them were unclear:
Add a Single Product block and select a product.
This was unclear if I should be selecting any product or the variable product I just created, or a variation of the product I just created. Turns out it was the variable product and not the variation, it may be helpful to mention this, as the UI makes it also look like that your forced to select a variation although the variable product is also an option.

Sorry to hear that :-(
Yes, the idea is to test the variable product you just created. So, you should select the product, not a variation. I’ve updated the testing instructions to clarify this.

> The buttons are now violet
Maybe explicitly say that the buttons should be violet in the page editor, as I was confused why they weren't violet within the preview, but only in the editor context.

The button shouldn’t be violet. I added the comment `To test the style fix, you have to:` to specify that we were reproducing the error, but I realize now that it wasn’t as clear as I thought.
Steps 13 to 16 are meant to reproduce the visual error. The `Add to cart` button shouldn’t change color after adding the `woocommerce` class. After this PR, the button remains unchanged, as it should.

louwie17 reviewed (approved):
Approving this as it tested well, although I will note as already mentioned in the comments that I am still weirded out about the css change to avoid the theme styling from getting through, as I thought that would be something expected.
